cmd/zed/main.go

```diff
@@ -17,37 +21,57 @@ import (
 	_ "github.com/brimdata/zed/cmd/zed/dev/zst/cut"
 	_ "github.com/brimdata/zed/cmd/zed/dev/zst/inspect"
 	_ "github.com/brimdata/zed/cmd/zed/dev/zst/read"
+	"github.com/brimdata/zed/cmd/zed/drop"
+	"github.com/brimdata/zed/cmd/zed/index"
+	initcmd "github.com/brimdata/zed/cmd/zed/init"
 	"github.com/brimdata/zed/cmd/zed/lake"
-	_ "github.com/brimdata/zed/cmd/zed/lake/auth"
-	_ "github.com/brimdata/zed/cmd/zed/lake/branch"
-	_ "github.com/brimdata/zed/cmd/zed/lake/create"
-	_ "github.com/brimdata/zed/cmd/zed/lake/delete"
-	_ "github.com/brimdata/zed/cmd/zed/lake/drop"
-	_ "github.com/brimdata/zed/cmd/zed/lake/find"
-	_ "github.com/brimdata/zed/cmd/zed/lake/index"
-	_ "github.com/brimdata/zed/cmd/zed/lake/init"
-	_ "github.com/brimdata/zed/cmd/zed/lake/load"
-	_ "github.com/brimdata/zed/cmd/zed/lake/log"
-	_ "github.com/brimdata/zed/cmd/zed/lake/ls"
-	_ "github.com/brimdata/zed/cmd/zed/lake/merge"
-	_ "github.com/brimdata/zed/cmd/zed/lake/query"
-	_ "github.com/brimdata/zed/cmd/zed/lake/rename"
-	_ "github.com/brimdata/zed/cmd/zed/lake/revert"
-	_ "github.com/brimdata/zed/cmd/zed/lake/serve"
-	_ "github.com/brimdata/zed/cmd/zed/lake/use"
-	_ "github.com/brimdata/zed/cmd/zed/lake/vacate"
+	"github.com/brimdata/zed/cmd/zed/load"
+	"github.com/brimdata/zed/cmd/zed/log"
+	"github.com/brimdata/zed/cmd/zed/ls"
+	"github.com/brimdata/zed/cmd/zed/merge"
 	"github.com/brimdata/zed/cmd/zed/query"
+	"github.com/brimdata/zed/cmd/zed/rename"
+	"github.com/brimdata/zed/cmd/zed/revert"
 	"github.com/brimdata/zed/cmd/zed/root"
+	"github.com/brimdata/zed/cmd/zed/serve"
+	"github.com/brimdata/zed/cmd/zed/use"
+	"github.com/brimdata/zed/cmd/zed/vacate"
+	"github.com/brimdata/zed/pkg/charm"
 )
 
 func main() {
-	zed := root.Zed
-	zed.Add(api.Cmd)
-	zed.Add(dev.Cmd)
-	zed.Add(query.Cmd)
-	zed.Add(lake.Cmd)
-	if err := zed.ExecRoot(os.Args[1:]); err != nil {
+	lakeCmd(auth.Cmd)
+	lakeCmd(branch.Cmd)
+	lakeCmd(create.Cmd)
+	lakeCmd(deletecmd.Cmd)
+	lakeCmd(drop.Cmd)
+	lakeCmd(index.Cmd)
+	lakeCmd(initcmd.Cmd)
+	lakeCmd(load.Cmd)
+	lakeCmd(log.Cmd)
+	lakeCmd(ls.Cmd)
+	lakeCmd(merge.Cmd)
+	lakeCmd(query.Cmd)
+	lakeCmd(rename.Cmd)
+	lakeCmd(revert.Cmd)
+	lakeCmd(serve.Cmd)
+	lakeCmd(use.Cmd)
+	lakeCmd(vacate.Cmd)
+	root.Zed.Add(dev.Cmd)
+	if err := root.Zed.ExecRoot(os.Args[1:]); err != nil {
 		fmt.Fprintf(os.Stderr, "%s\n", err)
 		os.Exit(1)
 	}
 }
+
+func lakeCmd(spec *charm.Spec) {
+	wrapnew := spec.New
+	spec.New = func(parent charm.Command, fs *flag.FlagSet) (charm.Command, error) {
+		lk, err := lake.New(parent.(*root.Command), fs)
+		if err != nil {
+			return nil, err
+		}
+		return wrapnew(lk, fs)
+	}
+	root.Zed.Add(spec)
+}
```

Comment:
This and `lake.New` don't feel quite right.

How about adding something like `type LakeFlag { Lake string }` to package `cli` (I know we already have `cli/lakeflags`, but we can rename that later) to handle `-lake` and giving it `Connection` and `Open` methods (to take the place of the corresponding methods in `cmd/zed/lake/command.go`)?

Reply:
What command would use `type LakeFlag`, root.Command?

Reply:
Should `LakeFlag` be present in the `dev` sub-commands?

Reply:
No, `LakeFlag` would go into any command that accesses a lake (i.e., the same commands passed to `lakeCmd` here).

Reply:
Oh I see. I kind of like that!

Reply:
Would need to also add ConfigDir to LakeFlag (along with the credential access methods) but this simplifies things nicely!